File: apps/macos/Sources/WispProjectBrowserUI/ProjectBrowserModel.swift
import AppKit
import SwiftUI
import WispProjectBrowser

@MainActor
public final class ProjectBrowserModel: ObservableObject {
    @Published private(set) var projects: [ProjectSummary] = []
    @Published private(set) var recentSessions: [BrowserSession] = []
    @Published private(set) var sessions: [BrowserSession] = []
    @Published private(set) var activeProjectID: String?
    @Published private(set) var activeSessionID: String?
    @Published private(set) var messages: [BrowserMessage] = []
    @Published private(set) var transcriptLoading = false
    @Published private(set) var nextBeforeSeq: Int64?
    private var transcriptGeneration = UUID()
    @Published private(set) var sessionsLoading = false
    @Published private(set) var sessionError: String?
    private var navigationGeneration = UUID()
    @Published public private(set) var isLoading = false
    @Published private(set) var error: String?
    @Published private(set) var lastLoaded: Date?
    @Published private(set) var databaseURL: URL
    private let client: any ProjectBrowserQuerying

    public init() {
        let environment = ProcessInfo.processInfo.environment
        let support = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
        let defaultDatabase = support.appendingPathComponent("science.wisp-science/wisp-science/wisp.sqlite")
        let saved = environment["WISP_BROWSER_DATABASE"] ?? UserDefaults.standard.string(forKey: "projectBrowser.database")
        databaseURL = saved.map { URL(fileURLWithPath: $0) } ?? defaultDatabase
        let executable = environment["WISP_SERVICE_PATH"].map { URL(fileURLWithPath: $0) }
            ?? Bundle.main.url(forAuxiliaryExecutable: "wisp-service")
            ?? Bundle.main.bundleURL.appendingPathComponent("wisp-service")
        client = ProjectBrowserClient(executableURL: executable)
    }

    init(client: any ProjectBrowserQuerying, databaseURL: URL) {
        self.client = client
        self.databaseURL = databaseURL
    }

    public func refresh() async {
        guard !isLoading else { return }
        isLoading = true
        error = nil
        defer { isLoading = false }
        do {
            let snapshot = try await client.listProjects(databaseURL: databaseURL)
            let recent = try await client.listSessions(databaseURL: databaseURL, projectID: nil)
            projects = snapshot.projects
            recentSessions = recent
            lastLoaded = Date()
            if let id = activeProjectID { await openProject(id, sessionID: activeSessionID) }
        } catch {
            self.error = error.localizedDescription
        }
    }

    public func chooseDatabase() {
        let panel = NSOpenPanel()
        panel.title = "选择 Wisp 数据库"
        panel.message = "仅查询已有的 wisp.sqlite，不修改数据或执行数据库升级。"
        panel.canChooseDirectories = false
        panel.allowsMultipleSelection = false
        panel.directoryURL = databaseURL.deletingLastPathComponent()
        guard panel.runModal() == .OK, let url = panel.url else { return }
        databaseURL = url
        UserDefaults.standard.set(url.path, forKey: "projectBrowser.database")
        goHome()
        projects = []
        recentSessions = []
        lastLoaded = nil
        Task { await refresh() }
    }

    func goHome() {
        navigationGeneration = UUID()
        transcriptGeneration = UUID()
        messages = []
        nextBeforeSeq = nil
        transcriptLoading = false
        activeProjectID = nil
        activeSessionID = nil
        sessions = []
        sessionError = nil
        sessionsLoading = false
    }

    func openProject(_ id: String, sessionID: String? = nil) async {
        let generation = UUID()
        navigationGeneration = generation
        transcriptGeneration = UUID()
        messages = []
        nextBeforeSeq = nil
        transcriptLoading = false
        activeProjectID = id
        activeSessionID = sessionID
        sessions = []
        sessionError = nil
        sessionsLoading = true
        do {
            let rows = try await client.listSessions(databaseURL: databaseURL, projectID: id)
            guard generation == navigationGeneration else { return }
            sessions = rows
            if let sessionID, !rows.contains(where: { $0.id == sessionID }) {
                sessionError = "这个会话已不存在，请刷新项目列表。"
            } else if let selected = sessionID ?? rows.first?.id {
                await openSession(selected)
            }
        } catch {
            guard generation == navigationGeneration else { return }
            sessionError = error.localizedDescription
        }
        if generation == navigationGeneration { sessionsLoading = false }
    }

    func openSession(_ id: String, older: Bool = false) async {
        guard let projectID = activeProjectID, sessions.contains(where: { $0.id == id }) else { return }
        if older && transcriptLoading { return }
        let generation = UUID()
        transcriptGeneration = generation
        activeSessionID = id
        let cursor = older ? nextBeforeSeq : nil
        if !older { messages = []; nextBeforeSeq = nil }
        transcriptLoading = true
        sessionError = nil
        do {
            let page = try await client.transcript(databaseURL: databaseURL, projectID: projectID, sessionID: id, beforeSeq: cursor)
            guard generation == transcriptGeneration else { return }
            messages = older ? page.messages + messages : page.messages
            nextBeforeSeq = page.nextBeforeSeq
        } catch {
            guard generation == transcriptGeneration else { return }
            sessionError = error.localizedDescription
        }
        if generation == transcriptGeneration { transcriptLoading = false }
    }

    func reveal(_ project: ProjectSummary) {
        let url = URL(fileURLWithPath: project.workspaceDirectory, isDirectory: true)
        guard Self.workspaceExists(project) else {
            error = "项目目录不存在或当前无法访问：\(project.workspaceDirectory)"
            return
        }
        NSWorkspace.shared.activateFileViewerSelecting([url])
    }

    static func workspaceExists(_ project: ProjectSummary) -> Bool {
        guard !project.workspaceDirectory.isEmpty else { return false }
        var directory: ObjCBool = false
        return FileManager.default.fileExists(atPath: project.workspaceDirectory, isDirectory: &directory)
            && directory.boolValue
    }
}

File: apps/windows/Wisp.ProjectBrowser/ProjectBrowserModel.cs
using Wisp.ProjectBrowser.Contracts;

namespace Wisp.ProjectBrowser;

/// <summary>UI-thread state with independent refresh, navigation and transcript generations.</summary>
public sealed class ProjectBrowserModel(IProjectBrowserClient client, string databasePath) : IDisposable
{
    public event Action? Changed;
    public string DatabasePath { get; private set; } = databasePath;
    public IReadOnlyList<ProjectSummary> Projects { get; private set; } = [];
    public IReadOnlyList<BrowserSession> RecentSessions { get; private set; } = [];
    public IReadOnlyList<BrowserSession> Sessions { get; private set; } = [];
    public IReadOnlyList<BrowserMessage> Messages { get; private set; } = [];
    public string? ActiveProjectId { get; private set; }
    public string? ActiveSessionId { get; private set; }
    public long? NextBeforeSeq { get; private set; }
    public bool Loading { get; private set; }
    public bool SessionsLoading { get; private set; }
    public bool TranscriptLoading { get; private set; }
    public string? Error { get; private set; }
    public string? SessionError { get; private set; }
    public DateTimeOffset? LastLoaded { get; private set; }
    private int refreshGeneration, navigationGeneration, transcriptGeneration;
    private readonly CancellationTokenSource lifetime = new();

    public async Task RefreshAsync()
    {
        if (Loading) return;
        var generation = ++refreshGeneration;
        var navigation = navigationGeneration;
        Loading = true;
        Error = null;
        Notify();
        try
        {
            var snapshot = await client.ListProjectsAsync(DatabasePath, lifetime.Token);
            var recent = await client.ListSessionsAsync(DatabasePath, cancellationToken: lifetime.Token);
            if (generation != refreshGeneration) return;
            Projects = snapshot.Projects;
            RecentSessions = recent.Take(5).ToArray();
            LastLoaded = DateTimeOffset.Now;
            if (navigation == navigationGeneration && ActiveProjectId is { } id)
            {
                if (Projects.Any(p => p.Id == id)) await OpenProjectAsync(id, ActiveSessionId);
                else GoHome();
            }
        }
        catch (Exception ex) when (ex is not OperationCanceledException)
        {
            if (generation == refreshGeneration) Error = ex.Message;
        }
        catch (OperationCanceledException) { }
        finally { if (generation == refreshGeneration) { Loading = false; Notify(); } }
    }

    public async Task SetStarredAsync(string projectId, bool starred)
    {
        if (Loading || !Projects.Any(p => p.Id == projectId)) return;
        var generation = ++refreshGeneration;
        Loading = true; Error = null; Notify();
        try
        {
            var snapshot = await client.SetProjectStarredAsync(DatabasePath, projectId, starred, lifetime.Token);
            if (generation == refreshGeneration) { Projects = snapshot.Projects; LastLoaded = DateTimeOffset.Now; }
        }
        catch (OperationCanceledException) { }
        catch (Exception ex) { if (generation == refreshGeneration) Error = ex.Message; }
        finally { if (generation == refreshGeneration) { Loading = false; Notify(); } }
    }

    public async Task ChangeDatabaseAsync(string path)
    {
        ++refreshGeneration;
        Loading = false;
        DatabasePath = path;
        Projects = [];
        RecentSessions = [];
        LastLoaded = null;
        Error = null;
        GoHome();
        await RefreshAsync();
    }

    public void GoHome()
    {
        ++navigationGeneration;
        ++transcriptGeneration;
        ActiveProjectId = ActiveSessionId = null;
        Sessions = [];
        Messages = [];
        NextBeforeSeq = null;
        SessionsLoading = TranscriptLoading = false;
        SessionError = null;
        Notify();
    }

    public async Task OpenProjectAsync(string id, string? sessionId = null)
    {
        if (!Projects.Any(p => p.Id == id)) return;
        var generation = ++navigationGeneration;
        ++transcriptGeneration;
        ActiveProjectId = id;
        ActiveSessionId = sessionId;
        Sessions = [];
        Messages = [];
        NextBeforeSeq = null;
        TranscriptLoading = false;
        SessionsLoading = true;
        SessionError = null;
        Notify();
        try
        {
            var rows = await client.ListSessionsAsync(DatabasePath, id, lifetime.Token);
            if (generation != navigationGeneration) return;
            Sessions = rows.Where(s => s.ProjectId == id).ToArray();
            if (sessionId != null && !Sessions.Any(s => s.Id == sessionId)) SessionError = "这个会话已不存在，请刷新项目列表。";
            else if ((sessionId ?? Sessions.FirstOrDefault()?.Id) is { } selected) await OpenSessionAsync(selected);
        }
        catch (Exception ex) when (ex is not OperationCanceledException)
        {
            if (generation == navigationGeneration) SessionError = ex.Message;
        }
        catch (OperationCanceledException) { }
        finally { if (generation == navigationGeneration) { SessionsLoading = false; Notify(); } }
    }

    public async Task OpenSessionAsync(string id, bool older = false)
    {
        if (ActiveProjectId is not { } projectId || !Sessions.Any(s => s.Id == id)) return;
        if (older && (TranscriptLoading || id != ActiveSessionId || NextBeforeSeq == null)) return;
        var generation = ++transcriptGeneration;
        var cursor = older ? NextBeforeSeq : null;
        ActiveSessionId = id;
        if (!older) { Messages = []; NextBeforeSeq = null; }
        TranscriptLoading = true;
        SessionError = null;
        Notify();
        try
        {
            var page = await client.GetTranscriptAsync(DatabasePath, projectId, id, cursor, lifetime.Token);
            if (generation != transcriptGeneration) return;
            Messages = (older ? page.Messages.Concat(Messages) : page.Messages).DistinctBy(m => m.Sequence).OrderBy(m => m.Sequence).ToArray();
            NextBeforeSeq = page.NextBeforeSeq;
        }
        catch (Exception ex) when (ex is not OperationCanceledException)
        {
            if (generation == transcriptGeneration) SessionError = ex.Message;
        }
        catch (OperationCanceledException) { }
        finally { if (generation == transcriptGeneration) { TranscriptLoading = false; Notify(); } }
    }

    public IEnumerable<SearchResult> Search(string query)
    {
        bool Matches(string value) => value.Contains(query.Trim(), StringComparison.OrdinalIgnoreCase);
        if (ActiveProjectId == null)
            foreach (var p in Projects.Where(p => Matches(p.Name) || Matches(p.Description) || Matches(p.WorkspaceDirectory)))
                yield return new(p.Name, p.WorkspaceDirectory, p.Id, null);
        foreach (var s in (ActiveProjectId == null ? RecentSessions : Sessions).Where(s => Matches(s.Title)))
            yield return new(s.Title, Projects.FirstOrDefault(p => p.Id == s.ProjectId)?.Name ?? "会话", s.ProjectId, s.Id);
    }

    private void Notify() { if (!lifetime.IsCancellationRequested) Changed?.Invoke(); }
    public void Dispose() { lifetime.Cancel(); lifetime.Dispose(); }
}

public sealed record SearchResult(string Title, string Detail, string ProjectId, string? SessionId);
